File: backend/src/services/category.service.ts
import { categoryProvider } from '../providers/categoryProvider.js';
import { CategoryRes } from '../schemas/category.js';
import { BaseListResType, BaseSearchRequestType } from '../schemas/search.js';

/**
 * Service Layer — Category
 * Maps raw Prisma rows to the API response shape.
 */
export const categoryService = {
  async getAll(request: BaseSearchRequestType): Promise<BaseListResType> {
    const limit = request.limit;
    const rows = await categoryProvider.findAll(request);
    const hasNextPage = rows.length > limit;
    const items = hasNextPage ? rows.slice(0, limit) : rows;
    const nextCursor = hasNextPage ? items[items.length - 1].id : null;

    return {
      data: items.map((c) => CategoryRes.parse(c)),
      limit,
      nextCursor,
      hasNextPage,
    };
  },

  async findById(id: number) {
    const row = await categoryProvider.findById(id);
    return CategoryRes.parse(row);
  },

  async create(name: string) {
    return categoryProvider.create({ name });
  },

  async update(name: string, id: number) {
    return categoryProvider.update({ name }, id);
  },

  async delete(id: number) {
    return categoryProvider.delete(id);
  },

  async updateImg(id: number, img: string) {
    return categoryProvider.update({ img }, id);
  },
};

File: backend/app/services/category_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from app.providers.category_provider import category_provider
from app.utils.app_error import AppError
from typing import Optional

async def get_all(db: AsyncSession, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None) -> dict:
    rows = await category_provider.find_all(db, limit=limit, cursor=cursor, order=order, search=search)
    has_next = len(rows) > limit
    items = rows[:limit] if has_next else rows
    next_cursor = items[-1].id if has_next and items else None
    return {"data": items, "limit": limit, "nextCursor": next_cursor, "hasNextPage": has_next}

async def find_by_id(db: AsyncSession, category_id: int):
    cat = await category_provider.find_by_id(db, category_id)
    if not cat:
        raise AppError(404, "Category not found")
    return cat

async def create(db: AsyncSession, name: str):
    return await category_provider.create(db, name)

async def update(db: AsyncSession, name: str, category_id: int):
    cat = await category_provider.update(db, category_id, name=name)
    if not cat:
        raise AppError(404, "Category not found")
    return cat

async def update_img(db: AsyncSession, category_id: int, img: str):
    cat = await category_provider.update(db, category_id, img=img)
    if not cat:
        raise AppError(404, "Category not found")
    return cat

async def delete(db: AsyncSession, category_id: int):
    cat = await category_provider.soft_delete(db, category_id)
    if not cat:
        raise AppError(404, "Category not found")
    return cat
